remote build: fully preserve local sources and support packaging all sources

cjp256 opened the pull request:
Fetch all project parts and archive them inside of a git repo for
launchpad to build with.

- rewrite snapcraft.yaml shipped to launchpad to reflect source handling

- add yamlordereddictloader python dep to reserialize the updated
  snapcraft.yaml in a manner that looks similar to the original.

- drop Repo & Worktree classes in favor of a simpler launchpad-specific
  implementation.

- added a simple ProjectSourceCache class to manage persistent cache
  of a project's sources between runs.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

sergiusens commented on requirements.txt:
```diff
@@ -27,3 +27,4 @@ https://files.pythonhosted.org/packages/16/4a/b085a04d6dad79aa5c00c65c9b2bbcb2c6
 lxml==4.3.0; sys_platform == 'win32'
 GitPython==2.1.11
 launchpadlib==1.10.6
+yamlordereddictloader==0.4.0
```
why this and not enhance snapcraft.yaml_utils?

cjp256 replied:
Oh that would be much more appropriate :)

sergiusens commented on snapcraft/internal/remote_build/_launchpad.py:
```diff
@@ -246,3 +264,87 @@ def _download_file(self, url: str, name: str) -> None:
+    def _create_source_archive(self, source_dir: str, archive_path: str):
+        """Archive sources at source_dir into archive_path."""
+        logger.debug("creating source archive: {}".format(archive_path))
+        command = ["tar", "cJf", archive_path, "-C", source_dir, "."]
```
so source will contain the original snap directory, which is desired, but how will it behave wrt git commands?

cjp256 replied:
Yeah, `snapcraft` itself appears to build fine (with its use of git-describe)

cjp256 replied:
My first spin of this was to extend the source classes to support mirroring.  It was a lot more optimal for build deltas, but I kept winding up with git-inside-of-git issues.

At first I just did the tarball hack for the obvious .git case, but reconsidered that any `local` sources may have other git-tracking-unfriendly files.  I don't currently know of any data loss issues with tarballing things up in this manner, but it's possible...

cjp256 commented:
Reworked to:
- only package `local` source types by default.  this should eliminate cross-platform issues in the default configuration.
- add `--package-all-sources` to pull and package all source types for the builder.
- implemented source selector support (and will package up all arch variants)
- eliminated source project cache and restored a semblance of worktree, decoupled from launchpad-specifics

cjp256 commented:
Added local plugins support if plugin is not in `/snap` (snapd)
Tested against figlet, multipass, nextcloud, katacontainers, snapcraft, and snapd.

`snapd` is currently erroring out on:
```
If you must store these files within the snap/ directory, move them to snap/local/, which is ignored by snapcraft.
DEPRECATED: The 'build' keyword has been replaced by 'override-build'
See http://snapcraft.io/docs/deprecation-notices/dn8 for more information.
Searching for local plugin for x-builddeb
DEPRECATED: The 'build' keyword has been replaced by 'override-build'
See http://snapcraft.io/docs/deprecation-notices/dn8 for more information.
Skipping pull fontconfig-bionic (already ran)
Skipping pull fontconfig-xenial (already ran)
Skipping pull libc6 (already ran)
Skipping pull liblzma5 (already ran)
Skipping pull snapd (already ran)
Skipping pull squashfs-tools (already ran)
Skipping pull xdelta3 (already ran)
Building fontconfig-bionic 
Traceback (most recent call last):
  File "./build-from-security.py", line 3, in <module>
    import apt
ImportError: No module named 'apt'
Failed to run 'build': Exit code was 1.
Build failed
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 268, in run
    self.build()
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 257, in build
    env=env)
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 104, in run_build_command
    return self.backend.run(args, env=full_env, **kwargs)
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/lxd.py", line 502, in run
    subprocess.check_call(cmd, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['lxc', 'exec', 'lp-xenial-amd64', '--env', 'LANG=C.UTF-8', '--env', 'SHELL=/bin/sh', '--env', 'SNAPCRAFT_BUILD_INFO=1', '--env', 'SNAPCRAFT_IMAGE_INFO={"build-request-id": "lp-50402220", "build-request-timestamp": "2019-07-22T14:20:38Z", "build_url": "https://launchpad.net/~cjp256/+snap/snapcraft-f1b0caf3e3cb4582857c41e998d09b08/+build/625075"}', '--env', 'SNAPCRAFT_BUILD_ENVIRONMENT=host', '--env', 'http_proxy=http://10.10.10.1:8222/', '--env', 'https_proxy=http://10.10.10.1:8222/', '--env', 'GIT_PROXY_COMMAND=/usr/local/bin/snap-git-proxy', '--', '/bin/sh', '-c', 'cd /build/snapcraft-f1b0caf3e3cb4582857c41e998d09b08 && linux64 snapcraft']' returned non-zero exit status 2
```

I wonder if python-apt should be in build-packages? Kicked off another clean build in an attempt to repro...

On another note, "git: version" ends up mismatching the expected value.

cjp256 commented:
Added support for `version: git` and `version-script`

sergiusens commented:
Please, let's not support version-script and tell people to use `snapcraftctl set-version` instead (through a DN which we should do if we haven't already).

sergiusens commented:
Can `add --package-all-sources to pull and package all source types for the builder.` be a separate commit?

sergiusens commented on snapcraft/internal/remote_build/_worktree.py:
```diff
@@ -14,100 +14,263 @@
+        return part_name
+
+    def _get_part_cache_dir(self, part_name: str, selector=None) -> str:
+        """Get source cache directory for part_name."""
```
can you add a docstring for what the selector is?

cjp256 replied:
> Can `add --package-all-sources to pull and package all source types for the builder.` be a separate commit?



> Can `add --package-all-sources to pull and package all source types for the builder.` be a separate commit?

Will do.

sergiusens commented on snapcraft/internal/remote_build/_worktree.py:
```diff
@@ -14,100 +14,263 @@
+        if isinstance(source, list):
+            # Grammar case.
+            part_yaml["source"] = []
+            for s in source:
```
can you add a comment here about needing to take into account advanced grammar and maybe a snippet on how it looks?

sergiusens replied:
Oh you do a bit above...

cjp256 replied:
Yeah I'll add some detail for good measure, it is a bit of an unusual construct.

sergiusens commented on snapcraft/internal/remote_build/_worktree.py:
```diff
@@ -14,100 +14,263 @@
+            source_commit=part.get("source-commit", None),
+        )
+
+    def _printable_part_name(self, part_name: str, selector=None) -> str:
```
this one does not need to be a class member

sergiusens commented on snapcraft/internal/remote_build/_worktree.py:
```diff
@@ -14,100 +14,263 @@
+        ):
+            version = self._snapcraft_yaml.get("version", None)
+            version_script = self._snapcraft_yaml.get("version-script", None)
+            version = _version.get_version(version, version_script)
```
version-script can be anything, it should just not be supported in this case

cjp256 replied:
Added error for remote build, opened PR to add DN #2639

sergiusens commented:
can you reassess some of the logger.info calls? some seem to be implementation detail user facing messages.

sergiusens commented on snapcraft/internal/remote_build/_worktree.py:
```diff
@@ -14,100 +14,263 @@
+            part_yaml["source-type"] = "tar"
+
+            # Strip source attributes no longer relevant.
+            for key in [
```
we should add a TODO to make this more structured and pick these up from the actual source handler or schema considering that we will be finding ourselves in a potential place where we add a new key and forget to pop it here.

cjp256 replied:
I changed it to instead just pop all keys that start with "source-", and then set the source attributes we dictate here (i.e. source-type).

cjp256 commented:
> Please, let's not support version-script and tell people to use `snapcraftctl set-version` instead (through a DN which we should do if we haven't already).

DN?

sergiusens commented:
@cjp256 Deprecation Notice

cjp256 commented:
> Please, let's not support version-script and tell people to use `snapcraftctl set-version` instead (through a DN which we should do if we haven't already).

Added DN for version-script:
https://github.com/snapcore/snapcraft/pull/2639

cjp256 commented:
@sergiusens Pushed updates which should address your prior review.  Left one comment open unresolved for you to verify.

- Cleaned up user-facing output, add new docstrings
- broke out commit for --package-all-sources
- rebased onto origin/remote-build after 3.7 update
- error out when using `version-script` with UnsupportedVersionScriptError
- error out when not using a supported base.

cjp256 commented:
Added/updated unit tests and addressed some static warnings.

sergiusens commented on snapcraft/cli/remote.py:
```diff
@@ -124,6 +111,9 @@ def remote_build(
 
     project = get_project()
 
+    if project.info.base not in ("core", "core16", "core18"):
```
instead of this check, check for `is None` or we will automatically block dispatching "bare" and "core20" builds when they are ready....

cjp256 replied:
Makes sense, updated.

sergiusens commented on snapcraft/cli/remote.py:
```diff
@@ -124,6 +111,9 @@ def remote_build(
     project = get_project()
 
+    if project.info.base not in ("core", "core16", "core18"):
+        raise errors.InvalidBaseError(base=project.info.base)
```
Maybe "requires" base?
Can you quickly try if the builder errors out appropriately if you try and do something like "base: core20"?

sergiusens replied:
Add a TODO to use Project.is_legacy, I will work on that.

cjp256 replied:
How about for an error: 
```
class UnsupportedBaseError(RemoteBuildBaseError):

    fmt = (
        "Remote build currently requires that the project uses bases.\n"
        "Please specify an appropriate `base` keyword and try again."
    )
```

cjp256 replied:
`base` is really unprocessed with project.info.  So whatever I put into base is considered valid here (e.g. core20, core21) until the builder itself determines it isn't valid (which you'll see with the other PR #2642).

I updated to use get_build_base() for the time being, I think that's most appropriate?

sergiusens commented on snapcraft/internal/remote_build/_launchpad.py:
```diff
@@ -132,6 +134,8 @@ def start_build(self, timeout: int = 5, attempts: int = 5) -> int:
             if builds.entries:
                 break
+
+            # TODO: there may actually be a real build error here!
```
this change seems unnecessary considering #2642

cjp256 replied:
Indeed, removed.

sergiusens commented on snapcraft/internal/remote_build/_worktree.py:
```diff
@@ -14,100 +14,321 @@
+        self._project = project
+
+        # Get snapcraft yaml (as OrderedDict).
+        self._snapcraft_config = (
```
why store this if you are saving project?
If it is to avoid a second call, you are doing it right below.

cjp256 replied:
It's really to avoid several calls to it on anything iterating over / working on parts as the `parts` info is not directly available in project (or .info) AFAICT.  I figured it made sense two grab two copies, one original `self._snapcraft_config` and one working copy to freely modify `self._preopared_snapcraft_config`.

I could replace the refs to `self._snapcraft_config` with `project.__raw_snapcraft`?  I was just trying to avoid use of internal properties.

cjp256 commented:
Pushed updates to hopefully address comments.

cjp256 commented:
Fixed launchpad test case.

cjp256 commented:
Build failures for various 16.04 images I expect are unrelated:
```
+ snapcraft stage
Traceback (most recent call last):
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/git/__init__.py", line 83, in <module>
    refresh()
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/git/__init__.py", line 73, in refresh
    if not Git.refresh(path=path):
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/git/cmd.py", line 290, in refresh
    raise ImportError(err)
ImportError: Bad git executable.
The git executable must be specified in one of the following ways:
    - be included in your $PATH
    - be set via $GIT_PYTHON_GIT_EXECUTABLE
    - explicitly set via git.refresh()
All git commands will error until this is rectified.
This initial warning can be silenced or aggravated in the future by setting the
$GIT_PYTHON_REFRESH environment variable. Use one of the following values:
    - quiet|q|silence|s|none|n|0: for no warning or exception
    - warn|w|warning|1: for a printed warning
    - error|e|raise|r|2: for a raised exception
Example:
    export GIT_PYTHON_REFRESH=quiet
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/snap/snapcraft/x1/bin/snapcraft", line 11, in <module>
    load_entry_point('snapcraft==3.7', 'console_scripts', 'snapcraft')()
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/pkg_resources/__init__.py", line 489, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2843, in load_entry_point
    return ep.load()
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2434, in load
    return self.resolve()
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2440, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/snapcraft/cli/__main__.py", line 23, in <module>
    from ._runner import run
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/snapcraft/cli/_runner.py", line 33, in <module>
    from .remote import remotecli
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/snapcraft/cli/remote.py", line 21, in <module>
    from snapcraft.internal.remote_build import WorkTree, LaunchpadClient, InfoFile, errors
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/snapcraft/internal/remote_build/__init__.py", line 17, in <module>
    from ._worktree import WorkTree  # noqa: F401
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/snapcraft/internal/remote_build/_worktree.py", line 18, in <module>
    import git
  File "/snap/snapcraft/x1/lib/python3.5/site-packages/git/__init__.py", line 85, in <module>
    raise ImportError('Failed to initialize: {0}'.format(exc))
ImportError: Failed to initialize: Bad git executable.
The git executable must be specified in one of the following ways:
    - be included in your $PATH
    - be set via $GIT_PYTHON_GIT_EXECUTABLE
    - explicitly set via git.refresh()
All git commands will error until this is rectified.
```

cjp256 commented:
Going to late-import git (as it was before I foolishly lost it).

cjp256 commented:
- Introduced a new error (GitNotFoundVersionError) specifically for git required for `version: git`.
- Late-import git
- Move all repo git operations into launchpad (as it is currently launchpad-specific and simplifies)
- Rename InvalidVersionGitError -> GitNotFoundProviderError (and align error messages)

sergiusens commented on snapcraft/internal/remote_build/_launchpad.py:
```diff
@@ -31,7 +30,9 @@
 from . import errors
 
 import snapcraft
+
```
what is this empty line here?

cjp256 commented:
> You seem to have forgotten to remove git from requirements.txt?

Hah! I was a bit perplexed earlier when I couldn't find "git" in the requirements.txt.  I wondered how that package got installed... now I know.  It's one of _those_ packages that use camelcase...

cjp256 commented:
Pushed fixes for above comments (spacing, GitPython in requirements.txt)
